src/circuit/gadget/utilities/cond_swap.rs

```diff
@@ -0,0 +1,319 @@
+use super::{copy, CellValue, UtilitiesInstructions};
+use halo2::{
+    circuit::{Chip, Layouter},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Permutation, Selector},
+    poly::Rotation,
+};
+use pasta_curves::arithmetic::FieldExt;
+use std::marker::PhantomData;
+
+pub trait CondSwapInstructions<F: FieldExt>: UtilitiesInstructions<F> {
+    /// Variable representing cell with a certain value in the circuit.
+    type Var;
+
+    #[allow(clippy::type_complexity)]
+    /// Given an input pair (x,y) and a `swap` boolean flag, return
+    /// (y,x) if `swap` is set, else (x,y) if `swap` is not set.
+    fn swap(
+        &self,
+        layouter: impl Layouter<F>,
+        pair: (
+            <Self as CondSwapInstructions<F>>::Var,
+            <Self as CondSwapInstructions<F>>::Var,
+        ),
+        swap: Option<bool>,
+    ) -> Result<
+        (
+            <Self as CondSwapInstructions<F>>::Var,
+            <Self as CondSwapInstructions<F>>::Var,
+        ),
+        Error,
+    >;
+}
+
+/// A chip implementing a conditional swap.
+#[derive(Clone, Debug)]
+pub struct CondSwapChip<F> {
+    config: CondSwapConfig,
+    _marker: PhantomData<F>,
+}
+
+impl<F: FieldExt> Chip<F> for CondSwapChip<F> {
+    type Config = CondSwapConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct CondSwapConfig {
+    pub q_swap: Selector,
+    pub x: Column<Advice>,
+    pub y: Column<Advice>,
+    pub x_swapped: Column<Advice>,
+    pub y_swapped: Column<Advice>,
```

Comment:
Nit: the use of `(x, y)` here might suggest these are coordinates. Maybe rename uniformly to `(a, b)`?